### PR TITLE
imagenet update

### PR DESCRIPTION
**Pull request by XiangruiCAI**

### CODE REVIEW

File: examples/imagenet/ilsvrc12.h [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+  void WriteMean(Tensor &mean, string path);
+  Tensor ReadMean(string path);
+
+ private:
+  /// size for resizing
+  const size_t kImageSize = 256;
+  const size_t kImageNBytes = 3 * kImageSize * kImageSize;
+  /// size for cropping
+  const size_t kCropSize = 227;
+  /// buffer to store train/test_file_size images
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: I think you may not need buff.
both reader and writer have buffer.
you can write images one by one, the writer would buffer them and dump to disk periodically.

File: examples/imagenet/ilsvrc12.h [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+      for (int k = 0; k < resized.channels(); k++)
+        data[k * kImageSize * kImageSize + i * kImageSize + j] = resized.at<cv::Vec3b>(i, j)[k];
+  image.CopyDataFromHostPtr<unsigned char>(data, kImageNBytes);
+  delete []data;
+
+  return image;
+}
+
+void ILSVRC::Write(string outfile, singa::io::Mode mode) {
+  BinFileWriter bfwriter;
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: make writer a class member?

File: examples/imagenet/ilsvrc12.h [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+  bfwriter.Open(path, kCreate);
+  bfwriter.Write(path, encoder->Encode(input));
+  bfwriter.Flush();
+  bfwriter.Close();
+}
+
+void ILSVRC::CreateTrainData(string image_list, string input_folder,
+                             string output_folder, size_t file_size = 12800) {
+  std::vector<std::pair<string, int>> file_list;
+  size_t *sum = new size_t[kImageNBytes];
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: typically type of mean is float.
it may not be a big problem if use unsigned char to store mean..

File: examples/imagenet/ilsvrc12.h [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+  int label;
+  std::ifstream image_list_file(image_list.c_str(), std::ios::in);
+  while (image_list_file >> image_file_name >> label)
+    file_list.push_back(std::make_pair(image_file_name, label));
+  LOG(INFO) << "Data Shuffling";
+  std::shuffle(file_list.begin(), file_list.end(),
+               std::default_random_engine());
+  LOG(INFO) << "Total number of training images is " << file_list.size();
+  num_train_images = file_list.size();
+  num_train_images = 12900;
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: don't forget to remove it

File: examples/imagenet/ilsvrc12.h [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+  if (file_size == 0) file_size = num_train_images;
+  for (size_t imageid = 0; imageid < num_train_images; imageid++) {
+    string path = input_folder + "/" + file_list[imageid].first;
+    Tensor image = ReadImage(path);
+    auto image_data = image.data<unsigned char>();
+    for (size_t i = 0; i < kImageNBytes; i++)
+      sum[i] += image_data[i];
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]:
```
sum[i] += static_cast<unsigned int>(image_data[i]);
```
could be more safe.

File: examples/imagenet/ilsvrc12.h [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+    label = file_list[imageid].second;
+    Tensor lb(Shape{1}, kInt);
+    lb.CopyDataFromHostPtr<int>(&label, 1);
+    std::vector<Tensor> input;
+    input.push_back(image);
+    input.push_back(lb);
+    string encode_str = encoder->Encode(input);
+    buff.push_back(encode_str);
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: remove buff.
```
if writer is null
  create writer
if imageid+1 % file_size ==0
  delete writer
  create wirte with a new file path
writer.write(encoded_str);
```

File: examples/imagenet/ilsvrc12.h [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+    string file = output_folder + "/train" +
+                  std::to_string(num_train_images / file_size + 1) + ".bin";
+    int count = buff.size();
+    Write(file, kCreate);
+    LOG(INFO) << "Write " << count << " images into " << file;
+  }
+  Tensor mean = Tensor(Shape{3, kImageSize, kImageSize}, kUChar);
+  unsigned char *mean_data = new unsigned char[kImageNBytes];
+  for (size_t i = 0 ; i < kImageNBytes; i++)
+    mean_data[i] = (unsigned char)(sum[i] / num_train_images);
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: typically type of mean is float.
it may not be a big problem if use unsigned char to store mean..

File: examples/imagenet/ilsvrc12.h [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+  bfreader.Open(path);
+  bfreader.Read(&key, &value);
+  auto ret = decoder->Decode(value);
+  bfreader.Close();
+  return ret[0];
+}
+
+std::pair<Tensor, Tensor> ILSVRC::LoadData(int flag, string file,
+                                           size_t read_size, size_t offset, string mean_path) {
+  BinFileReader bfreader;
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: make bfreader a member?
add a member for last_read_file;
```
if file != last_read_file
  create reader (file)
  last_read_file = file
read batchsize images and label.

File: examples/imagenet/imagenet.cc [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+  auto cuda = std::make_shared<CudaGPU>();
+  net.ToDevice(cuda);
+  SGD sgd;
+  OptimizerConf opt_conf;
+  opt_conf.set_momentum(0.9);
+  auto reg = opt_conf.mutable_regularizer();
+  reg->set_coefficient(0.005);
+  sgd.Setup(opt_conf);
+  sgd.SetLearningRateGenerator(
+      [lr](int step) { return lr * std::pow(0.1, (step / 100000)); });
```
Comment (nudles) [inline examples/imagenet/imagenet.cc]: rename step to epoch;

File: examples/imagenet/imagenet.cc [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+  size_t train_file_size = 12800;
+  if (pos != -1) train_file_size = atoi(argv[pos + 1]);
+
+  pos = singa::ArgPos(argc, argv, "-readsize");
+  size_t read_size = 32;
+  if (pos != -1) read_size = atoi(argv[pos + 1]);
+*/
+  LOG(INFO) << "Start training";
+  singa::Train(nEpoch, lr, train_image_list, train_image_folder, 
+      test_image_list, test_image_folder, bin_folder, train_file_size, read_size);
```
Comment (nudles) [inline examples/imagenet/imagenet.cc]: rename it to alexnet.cc;

File: examples/imagenet/ilsvrc12.h [hunk 11]
```diff
@@ -0,0 +1,409 @@
+/************************************************************
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: add a ilsrc12.cc to include a main function for create the bin files.
update cmake to add a new target for the main function.

File: examples/imagenet/imagenet.cc [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+  //                     train_file_size);
+  //data.CreateTestData(test_image_list, test_image_folder, bin_folder);
+  auto net = CreateNet();
+  auto cuda = std::make_shared<CudaGPU>();
+  net.ToDevice(cuda);
+  SGD sgd;
+  OptimizerConf opt_conf;
+  opt_conf.set_momentum(0.9);
+  auto reg = opt_conf.mutable_regularizer();
+  reg->set_coefficient(0.005);
```
Comment (nudles) [inline examples/imagenet/imagenet.cc]: 0.0005?

File: examples/imagenet/alexnet.cc [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+  conv->add_pad(pad);
+  conv->set_bias_term(true);
+
+  ParamSpec *wspec = conf.add_param();
+  wspec->set_name(name + "_weight");
+  auto wfill = wspec->mutable_filler();
+  wfill->set_type("Gaussian");
+  wfill->set_std(std);
+
+  ParamSpec *bspec = conf.add_param();
```
Comment (nudles) [inline examples/imagenet/alexnet.cc]: some layers' bias is initialized to 1.

File: CMakeLists.txt [hunk 14]
```diff
@@ -1,7 +1,7 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.6)
 
 PROJECT(singa)
-SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -O3")
```
Comment (nudles) [inline CMakeLists.txt]: Typically -O2 is used. -O3 may introduce some wired optimization techniques that are not reproducible or hard to debug.

File: CMakeLists.txt [hunk 15]
```diff
@@ -21,9 +21,11 @@ INCLUDE_DIRECTORIES(${SINGA_INCLUDE_DIR})
 OPTION(USE_CBLAS "Use CBlas libs" ON)
 OPTION(USE_CUDA "Use Cuda libs" ON)
 OPTION(USE_CUDNN "Use Cudnn libs" ON)
-OPTION(USE_OPENCV "Use opencv" OFF)
+OPTION(USE_OPENCV "Use opencv" ON)
```
Comment (nudles) [inline CMakeLists.txt]: turn off options for optional dependent libs.

File: examples/imagenet/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,16 @@
+INCLUDE_DIRECTORIES(${CMAKE_CURRENT_SOURCE_DIR})
+INCLUDE_DIRECTORIES(${CMAKE_BINARY_DIR}/include)
+
+IF(USE_CUDNN)
```
Comment (nudles) [inline examples/imagenet/CMakeLists.txt]: since cpp verion of conv and pooling are merged. (lrn and batchnorm will be merged soon).
no need to check USE_CUDNN.

File: examples/imagenet/alexnet.cc [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+  net.Add(new CudnnDropout(), GenDropoutConf("drop7", 0.5));
+  net.Add(new Dense(), GenDenseConf("ip8", 1000, 0.01, 1));
+
+  return net;
+}
+
+void TrainOneEpoch(FeedForwardNet &net, ILSVRC &data,
+                                      std::shared_ptr<Device> device, int epoch,
+                                      string bin_folder,
+                                      size_t num_train_images,
```
Comment (nudles) [inline examples/imagenet/alexnet.cc]: would it be better to pass in the total num of training files instead of num of images and images per file?

File: examples/imagenet/alexnet.cc [hunk 18]
```diff
@@ -0,0 +1,403 @@
...
+//      train_y.ToDevice(device);
+      ttr.Tick();
+      auto ret = net.TrainOnBatch(epoch, train_x, train_y);
+      train_time += ttr.Elapsed();
+      loss += ret.first;
+      metric += ret.second;
+      b++;
+      //LOG(INFO) << "batch " << b << ", loss: " << ret.first;
+    }
+    if (fno % 20 == 0) {
```
Comment (nudles) [inline examples/imagenet/alexnet.cc]: pls use num of iterations as the print frequency. 
the fno would change if the train_file_size changes.

File: examples/imagenet/alexnet.cc [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+      train_time = 0.0f;
+      b = 0;
+    }
+  }
+}
+
+void TestOneEpoch(FeedForwardNet &net, ILSVRC &data,
+                                     std::shared_ptr<Device> device, int epoch,
+                                     string bin_folder, size_t num_test_images,
+                                     size_t read_size,
```
Comment (nudles) [inline examples/imagenet/alexnet.cc]: replace all 'read_size' to 'batch_size'

File: examples/imagenet/alexnet.cc [hunk 20]
```diff
@@ -0,0 +1,403 @@
...
+      load_time += timer.Elapsed();
+      remain -= n_read;
+      if (remain < 0) break;
+      if (n_read < read_size) break;
+    }
+    //LOG(INFO) << "num of test images: " << remain;
+    test_x.CopyData(prefetch_x);
+    test_y.CopyData(prefetch_y);
+    timer.Tick();
+    th = data.AsyncLoadData(kEval, binfile, read_size, &prefetch_x, &prefetch_y, &n_read);
```
Comment (nudles) [inline examples/imagenet/alexnet.cc]: if all test images are load together in one call, then no need to use AsyncLoadData.

File: examples/imagenet/ilsvrc12.h [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+  while (image_list_file >> image_file_name >> label)
+    file_list.push_back(std::make_pair(image_file_name, label));
+  LOG(INFO) << "Data Shuffling";
+  std::shuffle(file_list.begin(), file_list.end(),
+               std::default_random_engine());
+  LOG(INFO) << "Total number of training images is " << file_list.size();
+  size_t num_train_images = file_list.size();
+  num_train_images = 12900;
+  if (file_size == 0) file_size = num_train_images;
+  for (size_t imageid = 0; imageid < num_train_images; imageid++) {
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: maybe we can use openmp to accelerate this for loop.

File: include/singa/core/device.h [hunk 22]
```diff
@@ -24,17 +24,34 @@
 #include <string>
```
Comment (nudles) [inline include/singa/core/device.h]: it looks like that you mixed with the PR for opencl?

File: examples/imagenet/ilsvrc12.h [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+    }
+    images.push_back(image);
+  }
+  //auto iot = timer2.Elapsed();
+  int nimg = images.size();
+  //LOG(INFO) << "num of images: " << nimg;
+  *n_read = nimg;
+
+  //Timer omptime;
+  int nProcessors = omp_get_max_threads();
```
Comment (nudles) [inline examples/imagenet/ilsvrc12.h]: nthreads<-- set by users
vector<std::thread*> threads;
for (int i = 1; i < nthreads; i++) {
  threads.push_back(new thread(decodeTransform, i, std::ref(images), std::ref(x), std::ref(y));
}
decodeTransform(0, images, x, y);
for(auto th: threads) {
  th->join();
  delete th;
}
